nexus/db-queries/src/db/datastore/physical_disk.rs

```diff
@@ -90,14 +105,16 @@ impl DataStore {
         disk: PhysicalDisk,
     ) -> CreateResult<PhysicalDisk> {
         let conn = &*self.pool_connection_authorized(&opctx).await?;
-        Self::physical_disk_insert_on_connection(&conn, opctx, disk).await
+        let disk = Self::physical_disk_insert_on_connection(&conn, opctx, disk)
+            .await?;
+        Ok(disk)
     }
 
     pub async fn physical_disk_insert_on_connection(
         conn: &async_bb8_diesel::Connection<db::DbConnection>,
         opctx: &OpContext,
         disk: PhysicalDisk,
-    ) -> CreateResult<PhysicalDisk> {
+    ) -> Result<PhysicalDisk, TransactionError<Error>> {
```

Comment:
So here's the tension I feel with errors in the datastore:

- Functions which send queries to Diesel return `diesel::result::Error` types.
- However, a lot of functions in Nexus use the `omicron_common::api::external::Error` type as an error type. This type is handy for propagation through HTTP, because it directly corresponds with HTTP status codes.

These aliases -- `CreateResult`, `ListResult`, etc, all assume the `omicron_common::api::external::Error` type.

When the goal is to propagate an error out through dropshot, this type works great. However, if we ever want to match on particular errors, using the "external error type" sucks.

One example of this is within the context of retryable transactions: cockroachdb / diesel return a specific serialization error when a transaction fails due to contention, but could be retried: https://github.com/oxidecomputer/async-bb8-diesel/blob/1b832c734e850fd190bcaffcd4f4b871e8454875/src/async_traits.rs#L34-L41

Back in Omicron - if we're acting on the database, and we get an arbitrary "diesel error", and we convert it to an "external error", we've lost this transaction retry information.

The `TransactionError` type is basically a helper here:

https://github.com/oxidecomputer/omicron/blob/78ad95be5f32d77fd2d2997d9e67e5d225a1c8e0/nexus/db-queries/src/db/error.rs#L14-L29

It is just "whatever custom error you wanted (maybe an external error?) or a diesel error directly".

By propagating this type around, functions can still return external errors directly, but they can *also* return diesel errors that can preserve retry information.

Reply:
FWIW, this is the reason that our transactions within Omicron return a `DieselError` -- so they know when they can retry. See: https://github.com/oxidecomputer/omicron/blob/49b76b1c51e85f3e939ff6ab6ef4b5ce2c09f84b/nexus/db-queries/src/transaction_retry.rs#L116

(This gets parsed in async-bb8-diesel, in https://github.com/oxidecomputer/async-bb8-diesel/blob/1b832c734e850fd190bcaffcd4f4b871e8454875/src/async_traits.rs#L270-L274 )

Reply:
Aha yes, that makes sense. Thanks.